mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp

```diff
@@ -2342,6 +2342,19 @@ ParseResult LLVMFuncOp::parse(OpAsmParser &parser, OperationState &result) {
   result.addAttribute(getFunctionTypeAttrName(result.name),
                       TypeAttr::get(type));
 
+  if (succeeded(parser.parseOptionalKeyword("vscale_range"))) {
```

Comment:
@Leporacanthicus  one small thing I forgot to mention before is that you can also print the attribute as part of the attribute dictionary, similar to the alignment:
`
  llvm.func @align_func() attributes {alignment = 2 : i64} {
    llvm.return
  }
`
That way you can save the code for the custom parsing and printing!

Reply:
I did try this, but I had some problems with getting a VScaleRangeAttr from the returned value for getAttribute - or something like that. I must say I didn't find it particularly easy to do what I wanted - probably because I've not really worked with MLIR attributes - especially not "other attributes that have something more than a simple value". And I though the `arm_streaming` attrribute was a good starting point... ;)

Reply:
I believe you can delete your printer and parser code and then the attribute should print and parse as part of the attribute dictionary. It is important to keep the attribute as a part of the operation argument list though:
`  OptionalAttr<UnnamedAddr>:$unnamed_addr,
    OptionalAttr<I64Attr>:$alignment,
    OptionalAttr<LLVM_VScaleRangeAttr>:$vscale_range
  );
`
This ensures the attribute is an inherent attribute of the operation itself, which means there is a convenient getter to access the attribute. Otherwise you indeed need to access the attribute using a key and cast it to the desired type.

However, the current solution works as well.

Reply:
Well, from what my experiments tell me, that won't work - if I have it there, tt needs to be an attribute of the function, that is read during parsing, because those are stored somewhere else than the ones given as "random attributes". 

Anyways, I will submit this on Moinday, unless I see any further comments that need fixing. I have technically already "gone home", but working from home, I have my computer in front of me after I've finished... :)